domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,174 @@
+  class PlayProgress(val type: PlayStatus, val position: Int, val duration: Int)
+  class MediaPlayerAlreadyInitException : Exception()
+  class MediaPlayerNotInitException : Exception()
+
+  private val mediaPlayer: MediaPlayer by lazy { MediaPlayer() }
+  private val playState = AudioMutableLiveData()
+  private var nextUpdateJob: Job? = null
+  private val seekBarLock = ReentrantLock()
+
+  private var prepared = false
+  private var observerActive = false
+  private var mediaPlayerActive = false
+
+  private val SEEKBAR_UPDATE_FREQUENCY = TimeUnit.SECONDS.toMillis(1)
+
+  /* Call function to begin loading a audio sourced specified by url
+  *  MediaPlayer must be inactive when calling this function */
+  fun initializeMediaPlayer(url: String) {
+    if (mediaPlayerActive) throw MediaPlayerAlreadyInitException()
+    mediaPlayerActive = true
+    setMediaPlayerListeners()
+    prepareDataSource(url)
+  }
+
+  /* Call function to change data source of media player
+  *  Puts media player in a preparing state */
+  fun changeDataSource(url: String) {
+    prepared = false
+    stopUpdatingSeekBar()
+    mediaPlayer.reset()
+    prepareDataSource(url)
+  }
+
+  private fun setMediaPlayerListeners() {
+    mediaPlayer.setOnCompletionListener {
+      stopUpdatingSeekBar()
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.COMPLETED, 0, mediaPlayer.duration)))
+    }
+    mediaPlayer.setOnPreparedListener {
+      prepared = true
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, it.duration)))
+    }
+  }
+
+  private fun prepareDataSource(url: String) {
+    mediaPlayer.setDataSource(context, Uri.parse(url))
+    mediaPlayer.prepareAsync()
+    playState.postValue(AsyncResult.pending())
+  }
+
+  /** Call function to play audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in paused state */
+  fun play() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (!mediaPlayer.isPlaying) {
+      mediaPlayer.start()
+      scheduleNextSeekBarUpdate()
+    }
+  }
+
+  /** Call function to pause audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in playing state */
+  fun pause() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (mediaPlayer.isPlaying) {
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PAUSED, mediaPlayer.currentPosition, mediaPlayer.duration)))
+      mediaPlayer.pause()
+      stopUpdatingSeekBar()
+    }
+  }
+
+  private fun scheduleNextSeekBarUpdate() {
+    if (observerActive && prepared) {
+      seekBarLock.withLock {
+        nextUpdateJob = CoroutineScope(backgroundDispatcher).launch {
+          delay(SEEKBAR_UPDATE_FREQUENCY)
+          updateSeekBar()
+          scheduleNextSeekBarUpdate()
+        }
+      }
+    }
+  }
+
+  private fun updateSeekBar() {
+    if (mediaPlayer.isPlaying) {
```

Comment:
Is it safe to access media player on a non-UI thread? We should at the very least be locking it here & everywhere since this can be accessed in a background dispatcher context.

Reply:
locked everywhere with shared variables